File: api-gateway/routes/audit.py
"""API Gateway — Audit log routes"""

import json
import logging
from datetime import datetime
from typing import Optional

from fastapi import APIRouter, HTTPException, Depends, Query
from pydantic import BaseModel

from api_gateway.services.db import get_connection
from api_gateway.services.auth import get_current_user

logger = logging.getLogger("api_gateway.audit")

router = APIRouter()


def _serialize(row) -> dict:
    """Make one audit row JSON-safe.

    Two columns come back as Python objects the response model cannot encode:
    ip_address is INET, which asyncpg hands over as an ipaddress.IPv4Address,
    and details is jsonb, which it hands over as a raw string. The first one
    raised PydanticSerializationError and took the whole endpoint down with a
    500 - invisibly, because the page renders a failed fetch as "no entries".
    """
    entry = dict(row)

    if entry.get("ip_address") is not None:
        entry["ip_address"] = str(entry["ip_address"])

    for key in ("id", "user_id", "resource_id"):
        if entry.get(key) is not None:
            entry[key] = str(entry[key])

    details = entry.get("details")
    if isinstance(details, str):
        try:
            entry["details"] = json.loads(details)
        except (TypeError, ValueError):
            pass

    return entry


def require_admin_or_manager(current_user: dict = Depends(get_current_user)) -> dict:
    """Dependency that requires admin or billing_manager role"""
    if current_user.get("role") not in ("admin", "billing_manager", "rcm_director"):
        raise HTTPException(status_code=403, detail="Manager+ access required")
    return current_user


# Who performed an entry. Three kinds of actor end up in this table and only
# the first has a users row:
#   * a person          -> user_id joins to users.username
#   * a sibling service -> user_id NULL, "service:ediparser" in details
#   * an unauthenticated caller -> user_id NULL, "anonymous" in details
# A user deleted since the entry was written also falls through to details,
# which is what keeps their history attributable.
ACTOR_SQL = "COALESCE(u.username, al.details->>'username', 'system')"


@router.get("/audit/actors", response_model=list[dict])
async def list_audit_actors(current_user = Depends(require_admin_or_manager)):
    """Everyone who appears in the log, with entry counts, for the filter.

    Driven by the data rather than the users table, so the dropdown can only
    ever offer a name that will actually match something - and so services and
    anonymous callers, which have no users row, are selectable too.
    """
    async with get_connection() as conn:
        rows = await conn.fetch(
            f"""
            SELECT {ACTOR_SQL} AS actor,
                   COUNT(*) AS entry_count,
                   MAX(al.created_at) AS last_seen
            FROM audit_log al
            LEFT JOIN users u ON u.id = al.user_id
            GROUP BY {ACTOR_SQL}
            ORDER BY COUNT(*) DESC, actor
            """
        )
        return [
            {
                "actor": r["actor"],
                "entry_count": r["entry_count"],
                "last_seen": r["last_seen"].isoformat() if r["last_seen"] else None,
            }
            for r in rows
        ]


@router.get("/audit", response_model=list[dict])
async def list_audit_log(
    action: Optional[str] = Query(None),
    resource_type: Optional[str] = Query(None),
    user_id: Optional[str] = Query(None),
    username: Optional[str] = Query(None, description="Actor name, including 'service:*' and 'anonymous'"),
    # Typed, so a query string becomes a datetime before it reaches asyncpg -
    # /audit?start_date=... returned 500 on the compliance log's own filter.
    start_date: Optional[datetime] = Query(None),
    end_date: Optional[datetime] = Query(None),
    limit: int = Query(200, ge=1, le=1000),
    offset: int = Query(0, ge=0),
    current_user = Depends(require_admin_or_manager),
):
    """List audit log entries (admin/manager only)"""
    async with get_connection() as conn:
        query = f"""
            SELECT al.id, al.user_id, u.username, al.action,
                   al.resource_type, al.resource_id, al.details,
                   al.ip_address, al.user_agent, al.created_at,
                   {ACTOR_SQL} AS actor
            FROM audit_log al
            LEFT JOIN users u ON u.id = al.user_id
        """
        params = []
        where_count = 0

        if action:
            query += f"{' WHERE' if where_count == 0 else ' AND'} al.action = ${where_count + 1}"
            params.append(action)
            where_count += 1

        if resource_type:
            query += f"{' WHERE' if where_count == 0 else ' AND'} al.resource_type = ${where_count + 1}"
            params.append(resource_type)
            where_count += 1

        if user_id:
            query += f"{' WHERE' if where_count == 0 else ' AND'} al.user_id = ${where_count + 1}"
            params.append(user_id)
            where_count += 1

        if username:
            # Filter on the same expression the dropdown is built from, so a
            # name offered by /audit/actors always returns its rows.
            query += f"{' WHERE' if where_count == 0 else ' AND'} {ACTOR_SQL} = ${where_count + 1}"
            params.append(username)
            where_count += 1

        if start_date:
            query += f"{' WHERE' if where_count == 0 else ' AND'} al.created_at >= ${where_count + 1}"
            params.append(start_date)
            where_count += 1

        if end_date:
            query += f"{' WHERE' if where_count == 0 else ' AND'} al.created_at <= ${where_count + 1}"
            params.append(end_date)
            where_count += 1

        query += f" ORDER BY al.created_at DESC LIMIT ${where_count + 1} OFFSET ${where_count + 2}"
        params.extend([limit, offset])

        rows = await conn.fetch(query, *params)
        return [_serialize(r) for r in rows]


@router.get("/audit/stats")
async def audit_stats(current_user = Depends(require_admin_or_manager)):
    """Get audit log summary statistics"""
    async with get_connection() as conn:
        actions = await conn.fetch(
            "SELECT action, COUNT(*) as cnt FROM audit_log GROUP BY action ORDER BY cnt DESC"
        )

        recent = await conn.fetchval(
            "SELECT COUNT(*) FROM audit_log WHERE created_at >= NOW() - INTERVAL '24 hours'"
        )

        logins = await conn.fetch(
            """SELECT u.username, u.role, COUNT(*) as login_count
               FROM audit_log al
               JOIN users u ON u.id = al.user_id
               WHERE al.action = 'login'
                 AND al.created_at >= NOW() - INTERVAL '7 days'
               GROUP BY u.username, u.role
               ORDER BY login_count DESC
               LIMIT 10"""
        )

        top_actions = [dict(a) for a in actions]
        top_logins = [dict(l) for l in logins]

        return {
            "total_entries": await conn.fetchval("SELECT COUNT(*) FROM audit_log"),
            "recent_24h": recent,
            "actions_breakdown": top_actions,
            "recent_logins": top_logins,
        }

File: crates/api-gateway/src/routes/audit.rs
//! Audit log routes.
//!
//! Ported from `api-gateway/routes/audit.py`. Read-only, manager-and-up (the
//! access-control middleware enforces `audit -> (MANAGER_UP, NOBODY)`), so
//! these handlers do no role checks of their own.

use axum::extract::{Query, State};
use axum::routing::get;
use axum::{Extension, Json, Router};
use chrono::{DateTime, Utc};
use denial_auth::rbac::Principal;
use denial_common::error::AppError;
use serde::Deserialize;
use sqlx::{QueryBuilder, Row};
use uuid::Uuid;

use crate::state::AppState;

/// Who performed an entry. Three kinds of actor end up in this table and only
/// the first has a `users` row: a person (`user_id` joins to `users`), a
/// sibling service (`user_id` NULL, `service:*` in details), and an
/// unauthenticated caller (`user_id` NULL, `anonymous` in details).
const ACTOR_SQL: &str = "COALESCE(u.username, al.details->>'username', 'system')";

#[derive(Deserialize)]
pub struct AuditQuery {
    pub action: Option<String>,
    pub resource_type: Option<String>,
    pub user_id: Option<String>,
    /// Actor name, including `service:*` and `anonymous`.
    pub username: Option<String>,
    pub start_date: Option<String>,
    pub end_date: Option<String>,
    #[serde(default = "default_limit")]
    pub limit: i64,
    #[serde(default)]
    pub offset: i64,
}

fn default_limit() -> i64 {
    200
}

fn organization_id(principal: &Principal) -> Result<Uuid, AppError> {
    principal
        .organization_id
        .as_deref()
        .and_then(|id| Uuid::parse_str(id).ok())
        .ok_or(AppError::Forbidden)
}

fn opt_string(row: &sqlx::postgres::PgRow, col: &str) -> Option<String> {
    row.try_get::<Option<String>, _>(col).ok().flatten()
}

fn opt_uuid_string(row: &sqlx::postgres::PgRow, col: &str) -> Option<String> {
    row.try_get::<Option<Uuid>, _>(col)
        .ok()
        .flatten()
        .map(|u| u.to_string())
}

fn opt_ts(row: &sqlx::postgres::PgRow, col: &str) -> Option<String> {
    row.try_get::<Option<DateTime<Utc>>, _>(col)
        .ok()
        .flatten()
        .map(|t| t.to_rfc3339())
}

/// Make one audit row JSON-safe. `ip_address` arrives already rendered to text
/// by `host()`, and `details` is decoded straight from `jsonb`.
fn serialize_row(row: &sqlx::postgres::PgRow) -> serde_json::Value {
    serde_json::json!({
        "id": opt_uuid_string(row, "id"),
        "user_id": opt_uuid_string(row, "user_id"),
        "username": opt_string(row, "username"),
        "action": opt_string(row, "action"),
        "resource_type": opt_string(row, "resource_type"),
        "resource_id": opt_uuid_string(row, "resource_id"),
        "details": row
            .try_get::<Option<serde_json::Value>, _>("details")
            .ok()
            .flatten()
            .unwrap_or(serde_json::Value::Null),
        "ip_address": opt_string(row, "ip_address"),
        "user_agent": opt_string(row, "user_agent"),
        "created_at": opt_ts(row, "created_at"),
        "actor": opt_string(row, "actor"),
    })
}

pub async fn list_audit_log(
    State(state): State<AppState>,
    Extension(principal): Extension<Principal>,
    Query(params): Query<AuditQuery>,
) -> Result<Json<Vec<serde_json::Value>>, AppError> {
    let organization_id = organization_id(&principal)?;
    let limit = params.limit.clamp(1, 1000);
    let offset = params.offset.max(0);

    let mut qb = QueryBuilder::<sqlx::Postgres>::new(
        "SELECT al.id, al.user_id, u.username, al.action, \
         al.resource_type, al.resource_id, al.details, \
         host(al.ip_address) AS ip_address, al.user_agent, al.created_at, ",
    );
    qb.push(ACTOR_SQL);
    qb.push(" AS actor FROM audit_log al LEFT JOIN users u ON u.id = al.user_id");

    qb.push(" WHERE al.organization_id = ")
        .push_bind(organization_id);
    let mut need_where = false;
    let prefix = |qb: &mut QueryBuilder<sqlx::Postgres>, need_where: &mut bool| {
        qb.push(if *need_where { " WHERE " } else { " AND " });
        *need_where = false;
    };

    if let Some(ref v) = params.action {
        prefix(&mut qb, &mut need_where);
        qb.push("al.action = ").push_bind(v);
    }
    if let Some(ref v) = params.resource_type {
        prefix(&mut qb, &mut need_where);
        qb.push("al.resource_type = ").push_bind(v);
    }
    if let Some(ref v) = params.user_id {
        let uid = Uuid::parse_str(v.trim())
            .map_err(|_| AppError::BadRequest("user_id must be a UUID".into()))?;
        prefix(&mut qb, &mut need_where);
        qb.push("al.user_id = ").push_bind(uid);
    }
    if let Some(ref v) = params.username {
        // Filter on the same expression the /audit/actors dropdown is built
        // from, so a name it offers always returns its rows.
        prefix(&mut qb, &mut need_where);
        qb.push(ACTOR_SQL).push(" = ").push_bind(v);
    }
    if let Some(ref v) = params.start_date {
        prefix(&mut qb, &mut need_where);
        qb.push("al.created_at >= ")
            .push_bind(v)
            .push("::timestamptz");
    }
    if let Some(ref v) = params.end_date {
        prefix(&mut qb, &mut need_where);
        qb.push("al.created_at <= ")
            .push_bind(v)
            .push("::timestamptz");
    }

    qb.push(" ORDER BY al.created_at DESC LIMIT ")
        .push_bind(limit)
        .push(" OFFSET ")
        .push_bind(offset);

    let rows = qb
        .build()
        .fetch_all(&state.pool)
        .await
        .map_err(AppError::Db)?;
    Ok(Json(rows.iter().map(serialize_row).collect()))
}

pub async fn list_audit_actors(
    State(state): State<AppState>,
    Extension(principal): Extension<Principal>,
) -> Result<Json<Vec<serde_json::Value>>, AppError> {
    let organization_id = organization_id(&principal)?;
    let sql = format!(
        "SELECT {ACTOR_SQL} AS actor, COUNT(*) AS entry_count, MAX(al.created_at) AS last_seen \
         FROM audit_log al LEFT JOIN users u ON u.id = al.user_id \
         WHERE al.organization_id = $1 \
         GROUP BY {ACTOR_SQL} ORDER BY COUNT(*) DESC, actor"
    );
    let rows = sqlx::query(&sql)
        .bind(organization_id)
        .fetch_all(&state.pool)
        .await
        .map_err(AppError::Db)?;
    let out = rows
        .iter()
        .map(|r| {
            serde_json::json!({
                "actor": opt_string(r, "actor"),
                "entry_count": r.try_get::<i64, _>("entry_count").unwrap_or(0),
                "last_seen": opt_ts(r, "last_seen"),
            })
        })
        .collect();
    Ok(Json(out))
}

pub async fn audit_stats(
    State(state): State<AppState>,
    Extension(principal): Extension<Principal>,
) -> Result<Json<serde_json::Value>, AppError> {
    let organization_id = organization_id(&principal)?;
    let actions = sqlx::query(
        "SELECT action, COUNT(*) AS cnt FROM audit_log WHERE organization_id = $1 GROUP BY action ORDER BY cnt DESC",
    )
    .bind(organization_id)
    .fetch_all(&state.pool)
    .await
    .map_err(AppError::Db)?;

    let recent_24h: (i64,) = sqlx::query_as(
        "SELECT COUNT(*) FROM audit_log WHERE organization_id = $1 AND created_at >= NOW() - INTERVAL '24 hours'",
    )
    .bind(organization_id)
    .fetch_one(&state.pool)
    .await
    .map_err(AppError::Db)?;

    let logins = sqlx::query(
        "SELECT u.username, u.role, COUNT(*) AS login_count \
         FROM audit_log al JOIN users u ON u.id = al.user_id \
         WHERE al.organization_id = $1 AND al.action = 'login' AND al.created_at >= NOW() - INTERVAL '7 days' \
         GROUP BY u.username, u.role ORDER BY login_count DESC LIMIT 10",
    )
    .bind(organization_id)
    .fetch_all(&state.pool)
    .await
    .map_err(AppError::Db)?;

    let total: (i64,) = sqlx::query_as("SELECT COUNT(*) FROM audit_log WHERE organization_id = $1")
        .bind(organization_id)
        .fetch_one(&state.pool)
        .await
        .map_err(AppError::Db)?;

    let actions_breakdown: Vec<serde_json::Value> = actions
        .iter()
        .map(|r| {
            serde_json::json!({
                "action": opt_string(r, "action"),
                "cnt": r.try_get::<i64, _>("cnt").unwrap_or(0),
            })
        })
        .collect();

    let recent_logins: Vec<serde_json::Value> = logins
        .iter()
        .map(|r| {
            serde_json::json!({
                "username": opt_string(r, "username"),
                "role": opt_string(r, "role"),
                "login_count": r.try_get::<i64, _>("login_count").unwrap_or(0),
            })
        })
        .collect();

    Ok(Json(serde_json::json!({
        "total_entries": total.0,
        "recent_24h": recent_24h.0,
        "actions_breakdown": actions_breakdown,
        "recent_logins": recent_logins,
    })))
}

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/", get(list_audit_log))
        .route("/actors", get(list_audit_actors))
        .route("/stats", get(audit_stats))
}
